### PR TITLE
Portability

### PR DESCRIPTION
**Pull request by poem2018**
code from ruike.
I put the random split training code in src/models folder, and the onnx and PyTorch-related code in src/utils folder. All the saved models will be saved under models/trained_models folder and the notebook will be put under ./notebook folder.

### CODE REVIEW

File: src/models/IN_dataGenerator_random.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
+from __future__ import print_function
```
Comment (jmduarte) [inline src/models/IN_dataGenerator_random.py]: what's the difference between this script and the `train_model.py` script? Can we merge the two into one?

File: src/utils/generate_train_test.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
+from __future__ import print_function
```
Comment (jmduarte) [inline src/utils/generate_train_test.py]: similarly, what's different about this script?

File: src/utils/pytorch_onnx_inference.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
+import glob
```
Comment (jmduarte) [inline src/utils/pytorch_onnx_inference.py]: can we merge this with `predict_model.py`? Maybe adding an option for whether to do inference with pytorch or both pytorch and onnx?

File: src/utils/tensorrt_inference.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
+import glob
```
Comment (jmduarte) [inline src/utils/tensorrt_inference.py]: maybe rename as `predict_model_tensorrt.py` or something to standardize it with the other scripts?

**Review by jmduarte — CHANGES_REQUESTED** [review]
Thanks looks good so far, just some questions

**Comment by poem2018** [comment]
Thanks for reviewing and suggestions.
The src/models/IN_dataGenerator_random.py<https://urldefense.com/v3/__https://github.com/FAIR4HEP/hbb_interaction_network/pull/9*discussion_r954400105__;Iw!!DZ3fjg!7YM4yLVJ-5R8aquW_iHyOIuw240WCH-TndyzSUNaWJKHzyikmVWpPRKQKSRb8yTo6jYADRhhZDSWTbossfFGoej64ULp$> aimed to randomly split the train test data without the limitations of h5 file constraints. Because in the previous splitting we always select several files as test and the rest as training, and the data inside one h5 file will always be counted as test or train. The random split will first store all the data together, and shuffle and split them into train and test (this is what src/utils/generate_train_test.py<https://urldefense.com/v3/__https://github.com/FAIR4HEP/hbb_interaction_network/pull/9*discussion_r954400616__;Iw!!DZ3fjg!7YM4yLVJ-5R8aquW_iHyOIuw240WCH-TndyzSUNaWJKHzyikmVWpPRKQKSRb8yTo6jYADRhhZDSWTbossfFGoXas3TNr$> does), then IN_dataGenerator_random.py<https://urldefense.com/v3/__https://github.com/FAIR4HEP/hbb_interaction_network/pull/9*discussion_r954400105__;Iw!!DZ3fjg!7YM4yLVJ-5R8aquW_iHyOIuw240WCH-TndyzSUNaWJKHzyikmVWpPRKQKSRb8yTo6jYADRhhZDSWTbossfFGoej64ULp$> will load those split data for training. In this way data inside one h5 file can also be split into train and test.

I can merge IN_dataGenerator_random.py<https://urldefense.com/v3/__https://github.com/FAIR4HEP/hbb_interaction_network/pull/9*discussion_r954400105__;Iw!!DZ3fjg!7YM4yLVJ-5R8aquW_iHyOIuw240WCH-TndyzSUNaWJKHzyikmVWpPRKQKSRb8yTo6jYADRhhZDSWTbossfFGoej64ULp$> to train_model.py and set a flag to decide whether we need to do random split(split happens within h5 file) or normal split(split happens outside h5 file)

src/utils/pytorch_onnx_inference.py<https://urldefense.com/v3/__https://github.com/FAIR4HEP/hbb_interaction_network/pull/9*discussion_r954401016__;Iw!!DZ3fjg!7YM4yLVJ-5R8aquW_iHyOIuw240WCH-TndyzSUNaWJKHzyikmVWpPRKQKSRb8yTo6jYADRhhZDSWTbossfFGoUan1fc8$>, for this script I can also merge it to predict_model.py.

src/utils/tensorrt_inference.py<https://urldefense.com/v3/__https://github.com/FAIR4HEP/hbb_interaction_network/pull/9*discussion_r954401244__;Iw!!DZ3fjg!7YM4yLVJ-5R8aquW_iHyOIuw240WCH-TndyzSUNaWJKHzyikmVWpPRKQKSRb8yTo6jYADRhhZDSWTbossfFGodTa50w7$>, for this script, I will rename it to standardize with the other scripts.

________________________________

File: src/models/IN_dataGenerator_random.py (hunk 1 above)
Comment (poem2018) [reply src/models/IN_dataGenerator_random.py]: Thanks for reviewing and suggestions.

The `src/models/IN_dataGenerator_random.py` aimed to randomly split the train test data without the limitations of h5 file constraints. Because in the previous splitting we always select several files as test and the rest as training, and the data inside one h5 file will always be counted as test or train. The random split will first store all the data together, and shuffle and split them into train and test (this is what `src/utils/generate_train_test.py` does), then `IN_dataGenerator_random.py` will load those split data for training. In this way data inside one h5 file can also be split into train and test.

I can merge `IN_dataGenerator_random.py` to train_model.py and set a flag to decide whether we need to do random split(split happens within h5 file) or normal split(split happens outside h5 file). Also rename the `src/utils/generate_train_test.py` so that it can show more clear about it's function

File: src/utils/pytorch_onnx_inference.py (hunk 3 above)
Comment (poem2018) [reply src/utils/pytorch_onnx_inference.py]: Yeah, for this script I can merge it to predict_model.py

File: src/utils/tensorrt_inference.py (hunk 4 above)
Comment (poem2018) [reply src/utils/tensorrt_inference.py]: will rename it.

File: src/models/models.py [hunk 5]
```diff
@@ -182,4 +182,4 @@ def forward(self, x, y):
     def tmul(self, x, y):  # Takes (I * J * K)(K * L) -> I * J * L
         x_shape = x.size()
         y_shape = y.size()
-        return torch.mm(x.view(-1, x_shape[2]), y).view(-1, x_shape[1], y_shape[1])
+        return torch.mm(x.reshape(-1, x_shape[2]), y).reshape(-1, x_shape[1], y_shape[1])
```
Comment (jmduarte) [inline src/models/models.py]: what's the reason for this change? just out of curiousity

File: src/models/predict_model.py [hunk 6]
```diff
@@ -172,18 +177,60 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
     )
 
-    gnn.load_state_dict(torch.load("IN_training/gnn_new_DR0_best.pth"))
-    print(sum(p.numel() for p in gnn.parameters() if p.requires_grad))
-    softmax = torch.nn.Softmax(dim=1)
+    if set_onnx is False:
+        print("11")
+        gnn.load_state_dict(torch.load("../../models/trained_models/gnn_new_best.pth"))
+        print(sum(p.numel() for p in gnn.parameters() if p.requires_grad))
+        #         softmax = torchs.nn.Softmax(dim=1)
```
Comment (jmduarte) [inline src/models/predict_model.py]: remove commented code

File: src/models/predict_model.py [hunk 7]
```diff
@@ -172,18 +177,60 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
+
+        for j in tqdm.tqdm(range(0, target_test.shape[0], batch_size)):
+            dummy_input_1 = torch.from_numpy(test[j : j + batch_size]).cuda()
+            dummy_input_2 = torch.from_numpy(test_sv[j : j + batch_size]).cuda()
+            out_test = gnn(dummy_input_1, dummy_input_2)
+            #             print(np.shape(torch.from_numpy(test[j : j + batch_size])))
```
Comment (jmduarte) [inline src/models/predict_model.py]: remove commented code

File: src/models/predict_model.py [hunk 8]
```diff
@@ -172,18 +177,60 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
+            #             out_test = softmax(gnn(torch.from_numpy(test[j : j + batch_size]).cuda()))
+            out_test = out_test.cpu().data.numpy()
+            #             print(np.shape(out_test))
```
Comment (jmduarte) [inline src/models/predict_model.py]: remove commented code

File: src/models/predict_model.py [hunk 9]
```diff
@@ -172,18 +177,60 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
-    for j in tqdm.tqdm(range(0, target_test.shape[0], batch_size)):
-        out_test = softmax(gnn(torch.from_numpy(test[j : j + batch_size]).cuda()))
-        out_test = out_test.cpu().data.numpy()
-        if j == 0:
-            prediction = out_test
-        else:
-            prediction = np.concatenate((prediction, out_test), axis=0)
-        del out_test
+    else:
+        print("22")
```
Comment (jmduarte) [inline src/models/predict_model.py]: remove

File: src/models/predict_model.py [hunk 10]
```diff
@@ -172,18 +177,60 @@ def main(args, save_path="", evaluating_test=True):  # noqa: C901
...
+            dummy_input_2 = test_sv[i : i + batch_size]
+
+            # Load the ONNX model
+            model = onnx.load(model_path)
+
+            # Check that the IR is well formed
+            onnx.checker.check_model(model)
+
+            # Print a human readable representation of the graph
+            # print(onnx.helper.printable_graph(model.graph))
```
Comment (jmduarte) [inline src/models/predict_model.py:211]: remove commented code

File: src/models/train_model.py [hunk 11]
```diff
@@ -17,9 +17,15 @@
 from src.data.h5data import H5Data
 from src.models.models import GraphNet
 
+# import sys
```
Comment (jmduarte) [inline src/models/train_model.py:20]: remove commented code

File: src/models/train_model.py [hunk 12]
```diff
@@ -81,29 +88,60 @@ def main(args):  # noqa: C901
...
+            spectators_name="spectator_subgroup",
+        )
+        data_val.set_file_names(files_val)
+
+        n_val = data_val.count_data()
+        n_train = data_train.count_data()
+        print("val data:", n_val)
+        print("train data:", n_train)
+    else:
+        t_X1_tr = np.load("//grand/RAPINS/ruike/npy_data1/data_X1_tr.npy")
```
Comment (jmduarte) [inline src/models/train_model.py]: can we remove hardcoded paths?

File: src/utils/generate_train_test.py (hunk 2 above)
Comment (jmduarte) [inline src/utils/generate_train_test.py]: Do we still need this script? If not, can we remove it?

File: src/utils/pytorch2onnx.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+import torch
+
+sys.path.append("..")
+from models.models import GraphNet  # noqa: E402
+
+sv_branch = 1
+N = 60  # number of charged particles
+N_sv = 5  # number of SVs
+n_targets = 2  # number of classes
+save_path = "//grand/RAPINS/ruike/new_hbb/test/"
```
Comment (jmduarte) [inline src/utils/pytorch2onnx.py]: Can we remove hardcoded paths

File: src/utils/pytorch2onnx.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+test_3 = np.swapaxes(test_3, 1, 2)
+test_spec = np.swapaxes(test_spec, 1, 2)
+
+fj_pt = test_spec[:, 0, 0]
+fj_eta = test_spec[:, 1, 0]
+fj_sdmass = test_spec[:, 2, 0]
+print("before", test_2.shape)
+print("before", test_3.shape)
+
+no_undef = fj_pt > -999  # no cut
```
Comment (jmduarte) [inline src/utils/pytorch2onnx.py]: if no cuts are made, can we remove this entirely

File: src/utils/pytorch2onnx.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    (fj_sdmass > min_msd)
+    & (fj_sdmass < max_msd)
+    & (fj_eta > min_eta)
+    & (fj_eta < max_eta)
+    & (fj_pt > min_pt)
+    & (fj_pt < max_pt)
+    & no_undef
+]
+
+print("after", test_2.shape)
```
Comment (jmduarte) [inline src/utils/pytorch2onnx.py]: remove

**Review by jmduarte — CHANGES_REQUESTED** [review]

**Comment by jmduarte** [comment]
thanks a lot @poem2018 , i made some suggestions

File: src/models/models.py (hunk 5 above)
Comment (poem2018) [reply src/models/models.py]: I previously run with the `view` function, but it will give error with the onnx inference, the error message is like below, so I change it to `reshape` function:
`RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.`
maybe we can add a comment here for notice, like needing to change to reshape when using onnx to inference?

**Comment by poem2018** [comment]
>

**Comment by poem2018** [comment]
> thanks a lot @poem2018 , i made some suggestions

Thanks for the suggestions, I modified them now
